### ISSUE TITLE
ICollection.CopyTo argument validation is inconsistent or non-existent

### ISSUE THREAD

**Issue by hughbe** [dotnet/corefx#9724]
Posting now after RTM, as suggested once upon a time.

The method `ICollection.CopyTo(Array array, int index)` has a multitude of invalid inputs. Here are some examples

### `array.GetLowerBound(0) != 0`
Suggested throw type: `ArgumentException`
Problem if not checked: `IndexOutOfRangeException` (bug)
Compatability impact: minimal (lower bound is very rarely not 0)

### `array.Rank != 1`
Suggested throw type: `ArgumentException`
Problem if not checked: `IndexOutOfRangeException` (bug)
Compatability impact: slight (array could be multidimensional but this is unlikely to be caught)

### `array == null`
Suggested throw type: `ArgumentNullException`
Problem if not checked: `NullReferenceException` (bug)
Compatability impact: medium (null is a common argument passed incorrectly)

### `index < 0`
Suggested throw type: `ArgumentOutOfRangeException`
Problem if not checked: `IndexOutOfRangeException` (bug)
Compatability impact: medium (e.g. letting user type in a custom index and catching IORE)

### `array.Count - index < collection.Count`
Suggested throw type: `ArgumentException`
Problem if not checked: `IndexOutOfRangeException` (bug)
Compatability impact: medium (e.g. letting user type in a custom index and catching IORE)

These are all I can think of at the moment.

## Examples
Here are some examples that I've encountered in the past. I haven't gone searching but I'm certain there are some more
- Regex ICollections only check if `array == null` and perform no other checks
- System.Security.Cryptography ICollections do not check the lower bound
- Hashtable, Hashtable.Keys and Hashtable.Values does not check the lower bound

## Proposition
Many ICollections do not perform all of these checks. The majority check for `array == null` and `index < 0`. The builtin `System.Collections.Generic` types have these checks, but many ICollections offer custom `CopyTo` implementations
We should discuss the following:
- Do the benefits of validating arguments (fixing a buggy implementation) outweigh the potential compat effects?
- Should we allow a non-zero lower bound (e.g. like SortedList)?
- Should we establish a consistent ordering for argument validation? Should existing argument validation checks be reordered?

/cc @stephentoub and @bartonjs who suggested creating a discussion

**Comment by Clockwork-Muse** [comment]
> <h3>`index < 0`</h3>

> Suggested throw type:  `ArgumentOutOfRangeException`
> Problem if not checked:  `IndexOutOfRangeException`  (bug)
> Compatability impact: medium (e.g. letting user type in a custom index and catching IORE)

> <h3>`array.Count - index < collection.Count`</h3>

> Suggested throw type:  `ArgumentException`
> Problem if not checked:  `IndexOutOfRangeException`  (bug)
> Compatability impact: medium (e.g. letting user type in a custom index and catching IORE)

Personally, I'm in agreement with the way the group implementing Midori handled these kinds of things; `ArgumentException` (especially index-related, and a few others) **almost always** means _programmer_, not user, error.  If a user can type in a random index, that actually needs to be validated as part of I/O, not passed down to where it's used.

**Comment by ianhays** [comment]
This is mostly a duplicate of https://github.com/dotnet/corefx/issues/7280, though with different examples.

The big push now is compat with the full framework, so it's unlikely that these exceptions will be changed as they're technically going to be breaking changes. I tried pushing through a few similar exception changes but feedback was strongly favored towards compat over consistency, so I closed that one out.  I'm going to close this issue as well, but hopefully we can revisit it at some point.
